source/common/grpc/async_client_manager_impl.cc

```diff
@@ -84,8 +86,24 @@ GoogleAsyncClientFactoryImpl::GoogleAsyncClientFactoryImpl(
   // Check metadata for gRPC API compliance. Uppercase characters are lowered in the HeaderParser.
   // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md
   for (const auto& header : config.initial_metadata()) {
-    if (!validateGrpcHeaderChars(header.key()) || !validateGrpcHeaderChars(header.value())) {
-      throw EnvoyException("Illegal characters in gRPC initial metadata.");
+    // Validate key
+    if (!validateGrpcHeaderChars(header.key())) {
+      throw EnvoyException(fmt::format("Illegal characters in gRPC initial metadata header key: {}.",header.key()));
+    }
+
+    // Validate value
+    // Binary base64 encoded
+    if (::absl::EndsWith(header.key(), "-bin")) {
+      if (header.value().length() > 0 && Base64::decode(header.value()) == EMPTY_STRING) {
+        throw EnvoyException(fmt::format("Illegal base64 value for gRPC initial metadata header key: {}.",header.key()));
+      }
+    // ASCII header
+    } else {
+      for (auto ch : header.value()) {
+        if (!(absl::ascii_isascii(ch))) {
```

Comment:
I don't think this check is sufficient: absl::ascii_isascii(ch) returns true if ch < 128 (see here: http://docs.ros.org/en/kinetic/api/abseil_cpp/html/ascii_8h_source.html). GRPC spec? (https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) defines ascii char as one between %x20-%x7E.

Reply:
Ok, I can change the check